docs/rfc/0000-application-architecture.md

```diff
@@ -0,0 +1,503 @@
+request requires authentication. If it does, the initiator requests a signature
+from the `ssh-agent` over the concatenation of
+
+    ticket || payload
+
+and includes the response, encoded as per [RFC8032], in the `signature` field.
+
+The receiver of an RPC with a non-null `signature` header MUST verify the
+`signature` against its own public key. If signature verification fails, it
+responds with a `forbidden` error, regardless of whether the RPC actually
+requires authentication as per the application semantics.
+
+
+## Architecture
+
+With the prerequisites in place, we lay out the following architecture:
+
+```
+                                   +-----------------+
+                                   |                 |
+         +-------------------------+  electron main  +------------------------+
+         |                         |                 |                        |
+         |                         +--------+--------+                        |
+         |                                  |                                 |
+         |                      +-----------+----------+                      |
+         |                      |                      |                      |
+         |                      |                      |                      |
++--------+--------+    +--------+--------+    +--------+--------+    +--------+--------+   +----------------+
+|                 |    |                 |    |                 |    |                 |   |                |
+|  peer-to-peer   |    |     CLIaaS      |    |    ssh-agent    |    |      PubSub     +---+      gitd      |
+|                 |    |                 |    |                 |    |                 |   |                |
++--------+--------+    +-----------------+    +--------+--------+    +--------+--------+   +-------+--------+
+         |                                             |         \            |           /        |
+         |                                 +-----------+          +----------------------+         |
+         |                                 |                                  |           \        |
+         |                         +-------+--------+                         |            +-------+--------+
+         |                         |                |                         |            |                |
+         +-------------------------+      CLI       +-------------------------+            |      git       |
+                                   |                |                                      |                |
+                                   +----------------+                                      +----------------+
+```
+
+The boxes in the middle form our "service layer", while the outer boxes
+represent interactive applications. The different components are descibed in
+more detail in the forthcoming sections.
+
+
+### Storage
```

Comment:
Am I right in thinking that implementing #461 means that a seed node won't need access to a key?

Reply:
No, a peer-to-peer networked program requires access to the key for transport-layer security and refs signing.

Reply:
So is this RFC punting on how a seed might handle that? Or would it be something like setting the `ssh-agent` to only ask once?

Reply:
Unless told explicitly, the `ssh-agent` doesn't ask for anything -- `ssh-add` may, if the key on-disk is encrypted. The agent _protocol_ (well, the single call we actually need) is simple enough so a frontend to some cloud KMS could be put together quickly. By saying that we support the protocol, we don't have to deal with any clown provider JSON abomination tho.